sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs

```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+
+            // Throw if the encryption protocol on the message doesn't match the version that this client library
+            // understands
+            // and is able to decrypt.
+            if (EncryptionConstants.ENCRYPTION_PROTOCOL_V1 != encryptionData.EncryptionAgent.Protocol)
+            {
+                throw new ArgumentException(
+                    "Invalid Encryption Agent. This version of the client library does not understand the " +
+                        $"Encryption Agent set on the queue message: {encryptionData.EncryptionAgent}");
+            }
+
+            return encryptionData;
+        }
+
+        private byte[] LocalManualEncryption(byte[] data, byte[] key, byte[] iv)
+        {
+            using (var aesProvider = new AesCryptoServiceProvider() { Key = key, IV = iv })
+            using (var encryptor = aesProvider.CreateEncryptor())
+            using (var memStream = new MemoryStream())
+            using (var cryptoStream = new CryptoStream(memStream, encryptor, CryptoStreamMode.Write))
+            {
+                cryptoStream.Write(data, 0, data.Length);
+                cryptoStream.FlushFinalBlock();
+                return memStream.ToArray();
+            }
+        }
+
+        #endregion
+
+        [TestCase(16)] // a single cipher block
+        [TestCase(14)] // a single unalligned cipher block
+        [TestCase(Constants.KB)] // multiple blocks
+        [TestCase(Constants.KB - 4)] // multiple unalligned blocks
+        [LiveOnly]
+        public void Upload(long dataSize)
+        {
+            var data = GetRandomBuffer(dataSize);
+            var key = new MockKeyEncryptionKey();
+            using (this.GetEncryptedBlockBlobClient(out var blob, key))
+            {
+                // upload with encryption
+                blob.Upload(new MemoryStream(data));
+
+                // download without decrypting
+                var normalBlob = new BlobClient(blob.Uri, this.GetNewSharedKeyCredentials()).Download().Value;
+                var encryptedData = new byte[normalBlob.ContentLength];
+                normalBlob.Content.Read(encryptedData, 0, encryptedData.Length);
+
+                // encrypt original data manually for comparison
+                var encryptionMetadata = GetAndValidateEncryptionData(normalBlob.Properties.Metadata);
+                byte[] expectedEncryptedData = LocalManualEncryption(
+                    data,
+                    key.UnwrapKey(null, encryptionMetadata.WrappedContentKey.EncryptedKey).ToArray(),
+                    encryptionMetadata.ContentEncryptionIV);
+
+                // compare data
+                Assert.AreEqual(expectedEncryptedData, encryptedData);
+            }
+        }
+
+        [TestCase(16)] // a single cipher block
+        [TestCase(14)] // a single unalligned cipher block
+        [TestCase(Constants.KB)] // multiple blocks
+        [TestCase(Constants.KB - 4)] // multiple unalligned blocks
+        [LiveOnly]
```

Comment:
We need recorded tests too.  Sean can help you get that setup.

Reply:
The client functionality being tested securely generates keys using `System.Security.Cryptography.RNGCryptoServiceProvider`. I don't know if its possible to record that. Sean isn't either.